File: lab-web/src/api/wire.ts
/** WebSocket wire decode helpers, mirroring ``lab/wire.py``. */

export const PROTOCOL_VERSION = 4;
export const NEURAL_INT_SCALE = 1e4;
export const JOINT_INT_SCALE = 1e4;
export const MUSCLE_INT_SCALE = 1e4;
export const TOUCH_INT_SCALE = 1e6;
export const MM_INT_SCALE = 1e6;

export function decodeScaled(values: number[], scale: number): Float32Array {
  const out = new Float32Array(values.length);
  const inv = 1 / scale;
  for (let i = 0; i < values.length; i++) out[i] = values[i] * inv;
  return out;
}

export function decodeSegmentsMm(sm: number[]): Float32Array {
  const n = Math.floor(sm.length / 2);
  const out = new Float32Array(n * 2);
  const inv = 1 / MM_INT_SCALE;
  for (let i = 0; i < n; i++) {
    out[i * 2] = sm[i * 2] * inv;
    out[i * 2 + 1] = sm[i * 2 + 1] * inv;
  }
  return out;
}

export function decodeComMm(cm: number[]): [number, number] {
  return [cm[0] / MM_INT_SCALE, cm[1] / MM_INT_SCALE];
}

export function decodeFiredBits(b64: string, n: number): Uint8Array {
  const raw = atob(b64);
  const bytes = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) bytes[i] = raw.charCodeAt(i);
  const out = new Uint8Array(n);
  for (let i = 0; i < n; i++) out[i] = (bytes[i >> 3] >> (i & 7)) & 1;
  return out;
}

File: lab/wire.py
"""Compact wire encoding helpers for the lab WebSocket stream.

Kept separate from the demo server's inline helpers so the lab can evolve the
wire format independently (adds ``Bi``, ``Trefi``, ``ja``, ``jv``, ``tc``,
``ma``, ``nm01``, ``M0i``, ``M1i`` per-neuron M_vector components, ``fe``,
``z``) while reusing the same quantization rules.
"""

from __future__ import annotations

import base64
import math
from typing import Any, Iterable

PROTOCOL_VERSION = 6
WIRE_FLOAT_SIG_DIGITS = 10
WIRE_SEGMENT_SIG_DIGITS = 6
NEURAL_INT_SCALE = 1e4
JOINT_INT_SCALE = 1e4
MUSCLE_INT_SCALE = 1e4
TOUCH_INT_SCALE = 1e6


def wire_float(x: float, *, sig_digits: int = WIRE_FLOAT_SIG_DIGITS) -> float:
    xf = float(x)
    if not math.isfinite(xf):
        return xf
    return float(f"{xf:.{sig_digits}g}")


def scaled_int(values: Iterable[float], *, scale: float = NEURAL_INT_SCALE) -> list[int]:
    return [int(round(float(x) * scale)) for x in values]


def mm_to_nm_int(mm: float) -> int:
    return int(round(float(mm) * 1e6))


def pack_bits(flags: list[int]) -> str:
    if not flags:
        return ""
    n = len(flags)
    buf = bytearray((n + 7) // 8)
    for i, v in enumerate(flags):
        if v:
            buf[i >> 3] |= 1 << (i & 7)
    return base64.b64encode(bytes(buf)).decode("ascii")


def quantize(obj: Any, *, sig_digits: int = WIRE_FLOAT_SIG_DIGITS) -> Any:
    if isinstance(obj, float):
        return wire_float(obj, sig_digits=sig_digits)
    if isinstance(obj, bool):
        return obj
    if isinstance(obj, (list, tuple)):
        return [quantize(v, sig_digits=sig_digits) for v in obj]
    if isinstance(obj, dict):
        return {k: quantize(v, sig_digits=sig_digits) for k, v in obj.items()}
    return obj
